File: strategy/generator/exit/ma.py
from dataclasses import dataclass

from core.models.moving_average import MovingAverageType
from core.models.parameter import (
    CategoricalParameter,
    Parameter,
    RandomParameter,
    StaticParameter,
)
from core.models.source import SourceType
from strategy.generator.exit.base import Exit, ExitType


@dataclass(frozen=True)
class MaExit(Exit):
    type: ExitType = ExitType.Ma
    source_type: Parameter = StaticParameter(SourceType.CLOSE)
    ma: Parameter = CategoricalParameter(MovingAverageType)
    period: Parameter = RandomParameter(5.0, 13.0, 2.0)

File: ta_lib/strategies/exit/src/ma.rs
use base::prelude::*;
use core::prelude::*;
use indicator::{ma_indicator, MovingAverageType};
use timeseries::prelude::*;

pub struct MaExit {
    source_type: SourceType,
    ma: MovingAverageType,
    period: usize,
}

impl MaExit {
    pub fn new(source_type: SourceType, ma: MovingAverageType, period: f32) -> Self {
        Self {
            source_type,
            ma,
            period: period as usize,
        }
    }
}

impl Exit for MaExit {
    fn lookback(&self) -> usize {
        self.period
    }

    fn close(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let ma = ma_indicator(&self.ma, data, self.source_type, self.period);

        (data.close().cross_under(&ma), data.close().cross_over(&ma))
    }
}
